LSPS6: LSP meta-data. (#114)
BitcoinAndLightningLayerSpecs/lsp: LSPS6/README.md

# LSPS6

| Name    | `.well-known`      |
| ------- |--------------------|
| Version | 1                  |

> | Status  | For Implementation |

Reviewer: Initial Status should be [draft](https://github.com/BitcoinAndLightningLayerSpecs/lsp?tab=readme-ov-file#status-specification).

The 'client' is the API consumer.

## Well-Known URI

[Well-known URIs][] are a way to discover information about a service.
The LSP SHOULD serve a well-known URI `/.well-known/bitcoin` to allow for public key discovery.

```json
{
  "icon": "https://acme.com/icon.png",
  "logo": "https://acme.com/logo.png",
  "email": "support@acme.com",
  "nostr": "npub1w...s2mfp9",
  "twitter": "@acme",
  "telegram": "@acme",
  "lightning": {
    "mainnet": {
      "public_keys": [
        "0326e6...2833a3",
        "0305f5...2e4e45"
      ]
    },
    "testnet": {
      "public_keys": [
        "03f060...869c00"
      ]
    },
    ...
  }
}
```

- `icon` LSP MAY include the URL of its icon.
- `logo` LSP MAY include the URL of its logo.
- `email` LSP MAY include its support email.
- `nostr` LSP MAY include its nost public key.
- `twitter` LSP MAY include its Twitter handle.
- `telegram` LSP MAY include its Telegram handle.
- `lightning`
  - `mainnet` LSP SHOULD include its mainnet lightning node public keys.
  - `testnet` LSP MAY include its testnet lightning node public keys.
  - `...` LSP MAY include its public keys for other network types.

> [Well-known URIs]: https://datatracker.ietf.org/doc/html/rfc8615

Reviewer: I would like to have all data in the `lsps6.get_info` api. Having the data on the web and on the lightning node includes a redundancy that is unnecessary IMO.

Using a web server to provide spoof resistance is great. But I suggest just providing a list of nodes ids in `/.well-known/bitcoin`.

Author: removed the socials and redundancy.

## API

### lsps6.get_info

`lsps6.get_info` provides generic information about the LSP.

The client SHOULD use the data in `url` up until the [TLD][] and append `/.well-known/bitcoin`.
So assuming the URL is `https://acme.com/lsp`, the client MUST use `https://acme.com/.well-known/bitcoin`.
To confirm authority the client MUST crawl this URL and
match the public key from the LSP with the public key from the well-known URI.

[TLD]: https://en.wikipedia.org/wiki/Top-level_domain

Any information that is available in the response of `lsps6.get_info`
should override the information in the well-known URI as it's more authoritative.

If there is a piece of information unavailable,
the client SHOULD use the information from the well-known URI as a fallback.

**Request** No parameters needed.

**Response**

```JSON
{

>   "url": "https://acme.com/lsp",

Reviewer: Spoof resistance check via the url should be optional. Is there any flag that I can see that indicates that `/.well-known/bitcoin` should be available?

Author: Sure, I can add a flag.
When I think about how I would implement the client, I would just ignore such a flag and always crawl the URI.
The end result is the same no green checkmark for you.

Reviewer: Think about my poor mobile phone doing this and how many request it could safe :)

Author: Hihi, OK I see your point. I always forget the mobile phone use-case.
I guess it's a shortcut to no green checkmark ;-)

  "icon": "https://acme.com/lsp-icon.png",

>   "logo": "https://acme.com/lsp-logo.png",

Reviewer: What's the difference between icon and logo? Does it need to be in a specific format? Maybe a specific resolution?

  "email": "lsp-support@acme.com",

>   "nostr": "npub1w...s2mfp9",

Reviewer: Is there a standard for npubs you can reference?

Author: I changed it to: 

npub according to [NIP-19](https://github.com/nostr-protocol/nips/blob/master/19.md).

not sure if that is what you intended.

  "twitter": "@acme-lsp",

>   "telegram": "@acme-lsp"

Reviewer: Twitter handle with or without the `@`? Please specify concretely what the user needs to put into this field.

}
```

- `url` LSP SHOULD include the URL of its website.
- `icon` LSP MAY include the URL of its icon.
- `logo` LSP MAY include the URL of its logo.
- `email` LSP MAY include its support email.
- `nostr` LSP MAY include its nost public key.
- `twitter` LSP MAY include its Twitter handle.
- `telegram` LSP MAY include its Telegram handle.

> ### lsps6.get_lsp_info

Reviewer: I am not sure if this method should be in `lsps6` if this info should be in it's according LSPS like LSPS2.

`lsps6.get_lsp_info` provides generic information about an LSPS spec.

**Request**
```JSON
{
  "spec": "LSPS1"
}
```

**Response**

```JSON
{
  "short": "Short description",
  "long": "Longer description of the details"
}
```

> LSP MUST always respond even when there is no support for a spec. LSP MUST include the following fields

Reviewer: Why always respond? Why not define an "NOT_FOUND" or "UNSUPPORTED" error?

Author: I didn't want to duplicate logic in this LSPS with differentiating if you support an LSPS or not. You have LSPS1 for that? All this LSPS does is provide a way to confirm authenticity and optionally add meta-data?

Reviewer: One more reason to put such method in their specific LSPS :)

# LSPS6 LSP meta-data

| Name    | `meta-data` |
| ------- |---------------|
| Version | 1             |
| Status  | Draft         |

LSPS6 requires [BOLT8][] as a transport layer.

[BOLT8]: https://github.com/lightning/bolts/blob/master/08-transport.md

## Motivation

Not all information is available through regular LSPS messages.
In order to provide more transparency, LSPS6 provides the ability to deliver extra information.
This extra information can be in the form of a URL, icon, logo and/or a support email address.
We need to be conscious about information spoofing.

### Actors

The 'LSP' is the API provider, and acts as the server.
The 'client' is the API consumer.

## DNS Public Key Verification

[BIP353][] provides a way to discover payment instructions over DNS.
Analogous to this the LSP SHOULD create a DNS TXT records in the form of

> `lsps.acme.com. 3600 IN TXT "public_key:0326e6...2833a3"` for public key verification.

Reviewer: ```suggestion
`lsps6.acme.com. 3600 IN TXT "public_key:0326e6...2833a3"` for public key verification.
```

Should we use `lsps6` as a name here in cast an other lsps uses DNS entries?

Reviewer: Please us the full node id in this example and not `...` to make the spec crystal clear.

If the LSP has multiple lightning network nodes, then the LSP SHOULD create multiple corresponding TXT records.

> If the LSP has multiple domains which is wishes to reference to, then the LSP SHOULD create multiple corresponding TXT records for each domain.

Reviewer: NIT
```suggestion
If the LSP has multiple domains which it wishes to reference to, then the LSP SHOULD create multiple corresponding TXT records for each domain.
```

> In the given example `0326e6...2833a3` should be replaced with the lightning network nodes' public key of the LSP (in zbase32 format).

Reviewer: Same here, full node id

The LSP must replace `acme.com` in `lsps.acme.com.` with their own domain.
Any TXT record not starting with `public_key:` can be ignored by the client.

The response from [lsps6.get_info](#lsps6get_info) may contain URLs and/or an email address in it's response.
When the LSP indicates support for DNS Public Key Verification via the flag is_dns_verification_available then
the client MUST confirm authenticity by verify the public key(s) from the domain's DNS record(s)

> with the public key that sends the [BOLT8][] messages.

Reviewer: `MUST` confirm in every case? That's very strict. `SHOULD` confirm if needed maybe? Or in what use cases MUST it confirm? If I just want to open a channel, I don't need to verify the domain ownership.

The client MUST verify all distinct domains.

[BIP353]: https://github.com/bitcoin/bips/blob/master/bip-0353.mediawiki

## API

### lsps6.get_info

`lsps6.get_info` provides generic information about the LSP like icon, logo, etc. Omitted fields are considered as not supported/available.
Any provided URLs should be valid and reachable. So starting with `http(s)://`.

**Request** No parameters needed.

**Response**

```JSON
{
  "url": "https://acme.com/lsp",
  "is_dns_verification_available": true,
  "icon": "https://acme.com/lsp-icon.png",
  "logo": "https://acme.com/lsp-logo.png",
  "email": "lsp-support@acme.com"
}
```

> - `url` LSP SHOULD include the URL to its website with a maximum URL length of 64 characters.

Reviewer: 64 character? 256 maybe? Also, it MUST be URL encoded.

  The LSP can choose to include a URL to a specific page that provides more information about the LSP or a top level domain.

> - `is_dns_verification_available` is an **optional** boolean field, if present it's purpose is to indicate if the LSP supports DNS Public Key Verification.

Reviewer: Why optional? It MUST be set to true or false. If no validation is provided then it should just be false. Making it optional doesnt make sense.

- `icon` is an **optional** field, if present it contains the URL to an icon with a maximum URL length of 128 characters.
  Icon is a squarely shaped smaller representation of the logo.
  Often the icon is the logo without the text to fit in a square shape.
  The maximum dimensions of the icon is 400 x 400px.

>   The minimum dimensions of the icon is 20 x 20px.

Reviewer: Use MUST, SHOULD, or MAY here. `Optional`, `is squarely` or any attribute like this is not spec compliant. Almost the full PR has this issue currently. Always use MUST/SHOULD/MAY.

- `logo` is an **optional** field, if present it contains the URL to a logo with a maximum URL length of 128 characters.

>   The maximum dimensions of the logo is 400 x 400px.

Reviewer 2: Are "weird" dimensions allowed, like 33x29? Do logos and icons have to be square?

Author: I'm open to suggestions ;-)
for icon it does say `Icon is a squarely shaped smaller representation of the logo.`?

Reviewer: You are the UI expert here Kobe. What do you think is best in TORQ and in a UI in general?

  The minimum dimensions of the logo is 20 x 20px.

> - `email` is an **optional** field, if present it contains the support email address with a maximum length of 128 characters.

Reviewer: format requirements? Pretty sure there is a spec for email formats.

# LSPS6 LSP meta-data

| Name    | `meta-data` |
| ------- |---------------|
| Version | 1             |
| Status  | Draft         |

LSPS6 requires [BOLT8][] as a transport layer.

[BOLT8]: https://github.com/lightning/bolts/blob/master/08-transport.md

## Motivation

Sometimes the client wants to obtain extra information from the LSP.
There might be information that is not available through regular LSPS messages.
An example for such information could be the pricing strategy for LSPS1.
This extra information can be in the form of a URL or an email address.
We need a way to prevent spoofing.

### Actors

The 'LSP' is the API provider, and acts as the server.
The 'client' is the API consumer.

## DNS Public Key Verification

[BIP353][] provides a way to discover payment instructions over DNS.
Analogous to this the LSP SHOULD create a DNS TXT record in the form of
`lsps.acme.com. 3600 IN TXT "public_key:0326e6...2833a3"` for public key verification.

> [BIP353]: https://github.com/bitcoin/bips/blob/master/bip-0353.mediawiki

Reviewer: Elaborate more on this. How is the TXT record structured. What is it used for? What do the numbers/letters after `public_key` mean? Is `public_key` mandatory? Is the public key in zbase32 or hex format? In a spec like this, you describe everything so it is idiot proof. Is any domain name allowed here or does it need to start with `lsps`? How do I distinguish this from other TXT records? Can I have multiple BIP353 records for the same domain? According to BIP353, this isn't invalid.

Also, rationales are needed why you want to do this. Lot's of open questions!

## API

### lsps6.get_info

`lsps6.get_info` provides generic information about the LSP like icon, logo, etc.

> It is possible to confirm authority over DNS when there are URLs (or an email address) in the response.

Reviewer: ```suggestion
A client MAY confirm the authority of this node public key over DNS when there are URLs (or an email address) in the response.
```

- Is authority the correct word? Isn't it `authenticity`?
- Does the client need to verify all URLS? Shouldn't one be enough?

To confirm authority the client MUST verify the public key(s) from the domain's DNS record(s)
with the public key that sends the [Bolt8][] messages.

**Request** No parameters needed.

**Response**

```JSON
{
  "url": "https://acme.com/lsp",
  "dns": true,
  "icon": "https://acme.com/lsp-icon.png",
  "logo": "https://acme.com/lsp-logo.png",
  "email": "lsp-support@acme.com"
}
```

> - `url` LSP SHOULD include the URL of its website.

Reviewer: What format should this URL have? Is there a max length?

> - `dns` LSP MAY include a boolean to indicate if the LSP has DNS Public Key Verification.

Reviewer: Rename to something more expressive. `is_dns_verification_available`?

The default is that it does not support it.
- `icon` LSP MAY include the URL of its icon.
Icon is a squarely shaped smaller representation of the logo.

> Often the icon is the logo without the text to fit in a square shape.

Reviewer: What are the dimension of the logo? Min/max?

> - `logo` LSP MAY include the URL of its logo.

Reviewer: What's the difference between the icon and logo?

> - `email` LSP MAY include its support email.

Reviewer: What the format of the email? What do you set the email to if it is not included? Is it null?

Reviewer 2: I agree, I would prefer something like:

```Markdown
`icon` is an **optional** field, if present BLAH BLAH
```

or

```Markdown
`icon` is a **nullable** field.  If non-`null`, BLAH BLAH
```
